apisix/plugins/opa/helper.lua

```diff
@@ -45,16 +52,136 @@ local function build_http_request(conf, ctx)
 end
 
 
-function _M.build_opa_input(conf, ctx, subsystem)
-    local request = build_http_request(conf, ctx)
+local function _build_http_route(conf, ctx)
+    local route_id = ctx.route_id
+    local routes = get_routes()
+
+    for _, route in ipairs(routes) do
+        if route.value.id == route_id then
+            return core.table.deepcopy(route).value
+        end
+    end
+
+    return nil
+end
+
+
+local function build_http_route(conf, ctx, remove_upstream)
+    local route = _build_http_route(conf, ctx)
+
+    if remove_upstream and route and route.upstream then
+        route.upstream = nil
+    end
+
+    return route
+end
+
+
+local function _build_http_upstream(conf, ctx)
+    local route = build_http_route(conf, ctx, false)
+
+    if route then
+        if route.upstream then
+            return core.table.deepcopy(route.upstream)
+        else
+            local upstream_id = route.upstream_id
+            local upstreams = get_upstreams()
+
+            for _, upstream in ipairs(upstreams) do
+                if upstream.value.id == upstream_id then
+                    return core.table.deepcopy(upstream).value
+                end
+            end
+        end
+    end
+
+    return nil
+end
+
+
+local function build_http_upstream(conf, ctx)
+    local upstream = _build_http_upstream(conf, ctx)
+
+    if upstream and upstream.parent then
+        upstream.parent = nil
+    end
+
+    return upstream
+end
+
+
+local function _build_http_service(conf, ctx)
+    local service_id = ctx.service_id
+
+    -- possible that the route is not bind a service
+    if service_id then
+        local services = get_services()
```

Comment:
We can use:
https://github.com/apache/apisix/blob/3c0b374808903388f5facc081dbdb2db672c7c69/apisix/init.lua#L415

Reply:
changed